misc/bazel/csharp.bzl

```diff
@@ -0,0 +1,73 @@
+load("@rules_dotnet//dotnet:defs.bzl", "csharp_binary", "csharp_library", "csharp_test", "publish_binary")
+load("@rules_pkg//pkg:mappings.bzl", "strip_prefix")
+load("@semmle_code//:dist.bzl", "pack_zip")
+load("//:defs.bzl", "codeql_platform")
+
+TARGET_FRAMEWORK = "net8.0"
+
+def codeql_csharp_library(name, **kwargs):
+    nullable = kwargs.pop("nullable", "enable")
+    target_frameworks = kwargs.pop("target_frameworks", [TARGET_FRAMEWORK])
+    csharp_library(name = name, nullable = nullable, target_frameworks = target_frameworks, **kwargs)
+
+def codeql_xunit_test(name, **kwargs):
+    nullable = kwargs.pop("nullable", "enable")
+    target_frameworks = kwargs.pop("target_frameworks", [TARGET_FRAMEWORK])
+
+    srcs = kwargs.pop("srcs", []) + [
+        "//csharp/extractor/Testrunner:Testrunner.cs",
+    ]
+
+    deps = kwargs.pop("deps", []) + [
+        "@paket.main//xunit",
+        "@paket.main//xunit.runner.utility",
+    ]
+
+    tags = kwargs.pop("tags", []) + ["csharp"]
+
+    csharp_test(
+        name = name,
+        deps = deps,
+        srcs = srcs,
+        nullable = nullable,
+        target_frameworks = target_frameworks,
+        tags = tags,
+        **kwargs
+    )
+
+def codeql_csharp_binary(name, **kwargs):
+    nullable = kwargs.pop("nullable", "enable")
+    visibility = kwargs.pop("visibility", ["//visibility:public"])
+    resources = kwargs.pop("resources", [])
+
+    # the entry assembly always has the git info embedded
+    resources.append("@semmle_code//:git_info")
+
+    target_frameworks = kwargs.pop("target_frameworks", [TARGET_FRAMEWORK])
+    csharp_binary_target = "bin/" + name
+    publish_binary_target = "publish/" + name
+    csharp_binary(name = csharp_binary_target, nullable = nullable, target_frameworks = target_frameworks, resources = resources, visibility = visibility, **kwargs)
+    publish_binary(
+        name = publish_binary_target,
+        binary = csharp_binary_target,
+        self_contained = True,
+        target_framework = TARGET_FRAMEWORK,
+        runtime_identifier = select(
+            {
+                "@platforms//os:macos": "osx-x64",  # always force intel on macos for now, until we build universal binaries
+                "//conditions:default": "",
+            },
+        ),
+    )
+
+    # we need to declare individual zip targets for each binary, as `self_contained=True` means that every binary target
+    # contributes the same runtime files. Bazel (rightfully) complains about this, so we work around this by creating individual zip files,
+    # and using zipmerge to produce the final extractor-arch file. For overlapping files, zipmerge chooses just one.
```

Comment:
I would very much like to avoid forcing zipmerge for any of our packs, as we really would better avoid zipping/unzipping dances.

Is this a `rules_pkg` warning we're trying to avoid here? I would prefer patching `rules_pkg` in our registry override (or long-term, upstream an option to silence/allow overlapping files) than using zipmerge.

Reply:
Here's how it would look like.

If we were to upstream, I'd add a `suppress_duplicate_warning` flag to the pkg rules to add when we know what we're doing.

Reply:
Maybe another approach might be to hand-roll a `pkg_overlay` rule combining `PackageFilegroupInfo` providers so that duplicate entries are removed

Reply:
uh, sorry, went for it, thought it would take me less time, but here's what I came up with

https://github.com/github/codeql/pull/16544

sorry if I'm duplicating your work. I do agree it's much better.

Reply:
Nah, I started with the `AssemblyInfo` part luckily :)